Add model to keep device metadata

eitoball opened the pull request:
This PR is for https://github.com/Safecast/safecastapi/issues/348

eitoball commented:
It is half-baked, but can be merged.

matschaffer commented:
Thanks! I'd like to take a look at the fields in realtime and the sheets
rob posted. Maybe there are some preliminary columns we should add here as
well.

On Tue, Jan 31, 2017 at 10:07 Eito Katagiri <notifications@github.com>
wrote:

> It is half-baked, but can be merged.
>
> —
> You are receiving this because your review was requested.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Safecast/ingest/pull/8#issuecomment-276243027>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAACsnD4YNEpaSt0BJyPSN8FFHSyDJFOks5rXolDgaJpZM4LxMgE>
> .
>
-- 

-Mat

matschaffer.com

matschaffer commented on app/api/v1/devices.rb:
```diff
@@ -0,0 +1,44 @@
+        delete ':id' do
+          device = ::Device.find(params[:id])
+          device.destroy!
+        end
```
The list & get is fine for now. Let's leave out the destructive methods until we have some answer around security though.

It's all just out in the open so would hate to build up metadata then have some rogue request kill it.

eitoball replied:
Right. I disabled create, update, and delete operations.

matschaffer commented on app/serializers/device_serializer.rb:
```diff
@@ -0,0 +1,7 @@
+class DeviceSerializer < ActiveModel::Serializer
+  attributes *%i(id location location_name payload created_at updated_at)
+
+  def location
```
Not sure if device needs a location or not. I'm sure it will be useful to have a notion of general area (like you have with `location_name`), but since solarcast is a semi-mobile device, it's "location" will likely be wherever it most recently reported a measurement.

eitoball replied:
I see. I deleted this method. All data goes into payload, we don't lose any data.

matschaffer replied:
Looks like I was wrong on the location bit https://github.com/Safecast/safecastapi/issues/355 - Glad we're keeping that data handy.

eitoball replied:
Reverted.

matschaffer commented on spec/app/api/v1/devices_spec.rb:
```diff
@@ -0,0 +1,79 @@
+    it 'returns 201 Created' do
+      post '/v1/devices', { numeric_id: 74 }
+      expect(last_response.status).to eq(201)
+      p ::Device.all.to_a
```
test stdout

eitoball replied:
Oops. Deleted.

matschaffer commented on spec/app/api/v1/devices_spec.rb:
```diff
@@ -0,0 +1,79 @@
+
+      it 'return an empty list' do
+        expect(last_response.status).to eq(200)
+        parsed_response = JSON.parse(last_response.body)
```
Maybe this should use a `be_json_including`-type matcher too?

eitoball replied:
I tried, but couldn't. I guess that it is `body` is an array of devices.

matschaffer commented:
@Morred will leave this for a bit incase you have feedback. I'm good to merge otherwise

matschaffer commented:
@eitoball Just saw https://safecast.slack.com/archives/api/p1485953995001530 - can we ensure the device numeric id is unsigned before we merge/push this?

eitoball commented:
>  can we ensure the device numeric id is unsigned before we merge/push this?

For now, I changed column type of `numeric_id` to `bigint` which accepts between -(2 ^ 63) and (2 ^ 63) -1. Since requirement is `"2 ^ 32 unsigned int"`, I added validation in Rails side, but not in DB for now.

matschaffer commented:
@eitoball should change `measurements.device_id` storage to match

eitoball commented:
> should change measurements.device_id storage to match

Right. Thx.
